dali/kernels/common/join/tensor_join_gpu_impl_test.cu

```diff
@@ -0,0 +1,345 @@
+    tensor_join::ConcatenateTensors(out[i], make_cspan(in_tensors), axis);
+  }
+}
+
+template <typename T>
+struct TensorJoinGPUTest : public ::testing::Test {
+  void TestRawKernel() {
+    CUDAStream stream = CUDAStream::Create(true);
+    InitData(stream);
+    RunRef();
+    RunRawKernel(stream);
+    CheckResult(stream);
+  }
+
+  void TestFullKernel(int num_iter = 3) {
+    if (new_axis)
+      TestFullKernelImpl<true>(num_iter);
+    else
+      TestFullKernelImpl<false>(num_iter);
+  }
+
+  template <bool new_axis>
+  void TestFullKernelImpl(int num_iter) {
+    using Kernel = TensorJoinGPU<T, new_axis>;
+    CUDAStream stream = CUDAStream::Create(true);
+    KernelManager mgr;
+    mgr.Resize<Kernel>(1, 1);
+    KernelContext ctx;
+    ctx.gpu.stream = stream;
+
+    for (int iter = 0; iter < num_iter; iter++) {
+      InitData(stream);
+      RunRef();
+
+      KernelRequirements &req = mgr.Setup<Kernel>(0, ctx, [&](int index) {
+        return &in_shapes[index];
+      }, in_shapes.size(), axis);
+      ASSERT_EQ(req.output_shapes.size(), 1);
+      ASSERT_EQ(req.output_shapes[0], out_shape);
+      mgr.Run<Kernel>(0, 0, ctx, out.gpu(stream), make_cspan(in_gpu_ptrs));
+
+      CUDA_CALL(cudaStreamSynchronize(stream));
+      CheckResult(stream);
+    }
+  }
+
+  void RunRef() {
+    RefTLSJoin(ref.cpu(), make_cspan(in_cpu_ptrs), axis);
+  }
+
+  void CheckResult(cudaStream_t stream) {
+    Check(out.cpu(stream), ref.cpu(stream));
+  }
+
+  void RunRawKernel(cudaStream_t stream) {
+    vector<tensor_join::InputDesc<T>> input_descs(N * njoin);
+    vector<tensor_join::OutputDesc<T>> output_descs(N);
+
+    auto out_tlv = out.gpu(stream);
+    FillDescs(make_span(output_descs), make_span(input_descs),
+              out_tlv, make_cspan(in_gpu_ptrs), axis);
+
+
+    CUDAEvent start = CUDAEvent::CreateWithFlags(0);
+    CUDAEvent end = CUDAEvent::CreateWithFlags(0);
+
+    DeviceBuffer<tensor_join::InputDesc<T>> gpu_input_descs;
+    DeviceBuffer<tensor_join::OutputDesc<T>> gpu_output_descs;
+    gpu_input_descs.from_host(input_descs, stream);
+    gpu_output_descs.from_host(output_descs, stream);
+
+    dim3 grid(1024, N);
+    dim3 block(32 * 8);
+    CUDA_CALL(cudaEventRecord(start, stream));
+
+    JoinTensorsKernel<<<grid, block, 0, stream>>>(
+        gpu_output_descs.data(), gpu_input_descs.data(), njoin);
+
+    CUDA_CALL(cudaEventRecord(end, stream));
+    CUDA_CALL(cudaStreamSynchronize(stream));
+
+    float time = 0;
+    CUDA_CALL(cudaEventElapsedTime(&time, start, end));
+    time *= 1e+6;  // to nanoseconds
+    int64_t size = 2 * out_tlv.num_elements() * sizeof(T);  // 2x because in+out
+    std::cerr << "Throughput: " << size / time << " GB/s\n";
+  }
+
+  void InitData(cudaStream_t stream) {
+    GenerateShapes();
+    FillInputs(stream);
+    GenerateListPointers(stream);
+    out.reshape(out_shape);
+    ref.reshape(out_shape);
+    FillOutput(stream);
+  }
+
+  void FillOutput(cudaStream_t stream) {
```

Comment:
```suggestion
  void InitOutput(cudaStream_t stream) {
```
or 
```suggestion
  void ResetOutput(cudaStream_t stream) {
```
?

Reply:
Reset. Init suggests that it would also resize.

Reply:
Sounds good